client/allocrunner/taskrunner/logmon_hook_unix_test.go

```diff
@@ -86,3 +87,76 @@ func TestTaskRunner_LogmonHook_StartCrashStop(t *testing.T) {
 	// Running stop should shutdown logmon
 	require.NoError(t, hook.Stop(context.Background(), nil, nil))
 }
+
+// TestTaskRunner_LogmonHook_ShutdownMidStart simulates logmon crashing while the
+// Nomad client is calling Start() and asserts that we recover and spawn a new logmon.
+func TestTaskRunner_LogmonHook_ShutdownMidStart(t *testing.T) {
+	t.Parallel()
+
+	alloc := mock.BatchAlloc()
+	task := alloc.Job.TaskGroups[0].Tasks[0]
+
+	dir, err := ioutil.TempDir("", "nomadtest")
+	require.NoError(t, err)
+	defer func() {
+		require.NoError(t, os.RemoveAll(dir))
+	}()
+
+	hookConf := newLogMonHookConfig(task.Name, dir)
+	hook := newLogMonHook(hookConf, testlog.HCLogger(t))
+
+	req := interfaces.TaskPrestartRequest{
+		Task: task,
+	}
+	resp := interfaces.TaskPrestartResponse{}
+
+	// First start
+	require.NoError(t, hook.Prestart(context.Background(), &req, &resp))
+	defer hook.Stop(context.Background(), nil, nil)
+
+	origState := resp.State
+	origHookData := resp.State[logmonReattachKey]
+	require.NotEmpty(t, origHookData)
+
+	// Pluck PID out of reattach synthesize a crash
+	reattach := struct {
+		Pid int
+	}{}
+	require.NoError(t, json.Unmarshal([]byte(origHookData), &reattach))
+	pid := reattach.Pid
+	require.NotZero(t, pid)
+
+	proc, _ := os.FindProcess(pid)
+
+	// Assert logmon is running
+	require.NoError(t, proc.Signal(syscall.Signal(0)))
+
+	// SIGSTOP would freeze process without it being considered
+	// exited; so this causes process to be non-exited at beginning of call
+	// then we kill process while Start call is running
+	require.NoError(t, proc.Signal(syscall.SIGSTOP))
+	// sleep for the signal to take effect
+	time.Sleep(1 * time.Second)
```

Comment:
How can you guarantee that the sleep is sufficient? Wondering if this will be flaky in CI

Reply:
I don't know of a good way to test if a process is stopped - unlike killing.  If it's flaky in CI, we can react - but considering that we will overhaul logmon plugin method to be inline with other plugin clients and this code would probably change, I'd avoid overengineering and react to CI problems as they come.

Reply:
So I take that back :).  ended up using `gopsutil.Process` utility to verify process sleep status in https://github.com/hashicorp/nomad/pull/5616/commits/658a734912e4355675729becb86dd149ac80b067!